pgjdbc/src/test/java/org/postgresql/test/ssl/SslTest.java

```diff
@@ -107,39 +109,51 @@ public static Iterable<Object[]> data() {
 
     Collection<Object[]> tests = new ArrayList<>();
 
-    for (SslMode sslMode : SslMode.VALUES) {
-      for (Hostname hostname : Hostname.values()) {
-        for (TestDatabase database : TestDatabase.VALUES) {
-          for (ClientCertificate clientCertificate : ClientCertificate.VALUES) {
-            for (ClientRootCertificate rootCertificate : ClientRootCertificate.VALUES) {
-              if ((sslMode == SslMode.DISABLE
-                  || database.rejectsSsl())
-                  && (clientCertificate != ClientCertificate.GOOD
-                  || rootCertificate != ClientRootCertificate.GOOD)) {
-                // When SSL is disabled, it does not make sense to verify "bad certificates"
-                // since certificates are NOT used in plaintext connections
-                continue;
-              }
-              if (database.rejectsSsl()
-                  && (sslMode.verifyCertificate()
-                      || hostname == Hostname.BAD)
-              ) {
-                // DB would reject SSL connection, so it makes no sense to test cases like verify-full
-                continue;
-              }
-              for (GSSEncMode gssEncMode : GSSEncMode.values()) {
-                if (gssEncMode == GSSEncMode.REQUIRE) {
-                  // TODO: support gss tests in /certdir/pg_hba.conf
+    for (SslNegotiation sslNegotiation :  SslNegotiation.values()) {
+      if (sslNegotiation == SslNegotiation.DIRECT) {
+        try (Connection con = TestUtil.openDB()) {
+          if (!TestUtil.haveMinimumServerVersion(con, ServerVersion.v17))
+            continue; // ignore direct connection unless we have version 17
+        } catch (SQLException e) {
+          fail("Failed to connect to the database: " + e.getMessage());
+        }
+      }
+      // iterate over all possible combinations of parameters
+      for (SslMode sslMode : SslMode.VALUES) {
```

Comment:
I suggest we skip `SslMode.DISABLE + SslNegotiation.DIRECT` combination as it is not different from `SslMode.DISABLE + SslNegotiation.POSTGRESQL`